### PR TITLE
[rpc] use callback func to do send & recv

### PR DESCRIPTION
**Pull request by yzhliu**
#3301 
@tqchen @eqy @t-vi

### CODE REVIEW

**Comment by tqchen** [comment]
let us keep the old setup for py(as it might gives better speed due to reduced ffi) and expose this other one

**Comment by yzhliu** [comment]
@tqchen OK.
any idea why cpu build is failing? seems other PRs are also failing due to same error.

**Comment by tqchen** [comment]
hmm, could due to inconsistency in cache to due updates introduced in the lastest python refactor, can you try to retrigger now that the PR has been merged?

**Comment by yzhliu** [comment]
@tqchen please review again

File: src/runtime/rpc/rpc_session.h [hunk 1]
```diff
@@ -253,6 +253,41 @@ class RPCSession {
   std::string remote_key_;
 };
 
+class CallbackChannel final : public RPCChannel {
+ public:
+  explicit CallbackChannel(PackedFunc fsend, PackedFunc frecv)
+      : fsend_(std::move(fsend)), frecv_(std::move(frecv)) {}
+
+  explicit CallbackChannel(PackedFunc fsend)
```
Comment (tqchen) [inline src/runtime/rpc/rpc_session.h]: Why is this constructor necessary? seems we can just remove this constructor

File: src/runtime/rpc/rpc_session.h [hunk 2]
```diff
@@ -253,6 +253,41 @@ class RPCSession {
   std::string remote_key_;
 };
 
+class CallbackChannel final : public RPCChannel {
```
Comment (tqchen) [inline src/runtime/rpc/rpc_session.h]: add a document about the CallbackChannel

**Comment by yzhliu** [comment]
@tqchen modified per comments.

**Comment by tqchen** [comment]
Thanks @yzhliu !

**Pull request merged**